File: hse/BPM161/pravilov/2/src/main/java/com/pravilov/hse/LockFreeSet.java
package com.pravilov.hse;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.atomic.AtomicMarkableReference;

public class LockFreeSet<T extends Comparable<T>> implements LockFreeSetInterface<T> {
    private Node head;
    private Node tail;

    public LockFreeSet() {
        head = new Node(null, ValueType.MIN);
        tail = new Node(null, ValueType.MAX);
        head.next.set(tail, false);
    }

    private PredCurrNodePair find(T value) {
        while (true) {
            Node pred = head;
            Node curr = pred.next.getReference();
            Node succ;

            while (true) {
                succ = curr.next.getReference();
                boolean cmk = curr.next.isMarked();
                if (cmk) {
                    if (!pred.next.compareAndSet(curr, succ, false, false)) {
                        break;
                    }
                    curr = succ;
                } else {
                    if (compareNodeWithValue(curr, value) >= 0) {
                        return new PredCurrNodePair(pred, curr);
                    }
                    pred = curr;
                    curr = succ;
                }
            }
        }
    }

    @Override
    public boolean add(T value) {
        while(true) {
            PredCurrNodePair predCurr = find(value);
            Node pred = predCurr.pred;
            Node curr = predCurr.curr;

            if (curr.value == value) {
                return false;
            } else {
                Node node = new Node(value);
                node.next.set(curr, false);
                if (pred.next.compareAndSet(curr, node, false, false)) {
                    return true;
                }
            }
        }
    }

    @Override
    public boolean remove(T value) {
        while (true) {
            PredCurrNodePair predCurr = find(value);
            Node pred = predCurr.pred;
            Node curr = predCurr.curr;

            if (compareNodeWithValue(curr, value) != 0) {
                return false;
            }
            Node succ = curr.next.getReference();
            if (!curr.next.compareAndSet(succ, succ, false, true)) {
                continue;
            }
            pred.next.compareAndSet(curr, succ, false, false);
            return true;
        }
    }

    @Override
    public boolean contains(T value) {
        Node curr = head;
        while (compareNodeWithValue(curr, value) < 0) {
            curr = curr.next.getReference();
        }
        return compareNodeWithValue(curr, value) == 0 && !curr.next.isMarked();
    }

    @Override
    public boolean isEmpty() {
        while (head.next.getReference() != tail) {
            if (!head.next.isMarked()) {
                return false;
            }
            Node curr = head.next.getReference();
            Node succ = curr.next.getReference();
            head.next.compareAndSet(curr, succ, false, false);
        }
        return true;
    }

    @Override
    public Iterator<T> iterator() {
        while (true) {
            List<Node> nodes1 = getListOfNodes();
            List<Node> nodes2 = getListOfNodes();
            if (nodes1.equals(nodes2)) {
                return nodes1.stream().map(node -> node.value).iterator();
            }
        }
    }

    private List<Node> getListOfNodes() {
        List<Node> nodes = new ArrayList<>();
        Node curr = head.next.getReference();
        while (curr != tail) {
            if (!curr.next.isMarked()) {
                nodes.add(curr);
            }
            curr = curr.next.getReference();
        }
        return nodes;
    }

    private class PredCurrNodePair {
        private Node pred;
        private Node curr;

        private PredCurrNodePair(Node pred, Node curr) {
            this.pred = pred;
            this.curr = curr;
        }
    }

    private enum ValueType {MIN, MAX, ORDINARY}

    private int compareNodeWithValue(Node node, T value) {
        if (node.valueType == ValueType.MIN) {
            return -1;
        }
        if (node.valueType == ValueType.MAX) {
            return 1;
        }
        return node.value.compareTo(value);
    }

    private class Node {
        private T value;
        private ValueType valueType;
        final private AtomicMarkableReference<Node> next = new AtomicMarkableReference<>(null, false);

        private Node(T value, ValueType valueType) {
            this.value = value;
            this.valueType = valueType;
        }

        private Node(T value) {
            this.value = value;
            this.valueType = ValueType.ORDINARY;
        }
    }
}

File: hse/BPM161/fedotov/2/src/main/kotlin/ru/hse/iisuslik/LockFreeSet.kt
package ru.hse.iisuslik

import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.atomic.AtomicStampedReference

/**
 * Lock-Free множество.
 * @param <T> Тип ключей
</T>
 */
class LockFreeSet<T : Comparable<T>> {

    private var currentVersion = AtomicInteger(0)
    private val head = Node(Node<T>(null, null), null)


    /**
     * Проверка множества на пустоту
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @return true если множество пусто, иначе - false
     */
    fun isEmpty(): Boolean {
        return makeSnapshot().isEmpty()
    }

    /**
     * Добавить ключ к множеству
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return false если value уже существует в множестве, true если элемент был добавлен
     */
    fun add(value: T): Boolean {
        while (true) {
            val (pred, curr) = find(value)
            if (curr.data != null && curr.data == value) {
                return false
            } else {
                val newNode = Node(null, value)
                newNode.next.set(curr, NOT_DELETED)
                val oldVersion = pred.version
                if (pred.next.compareAndSet(curr, newNode, oldVersion * 2, (oldVersion + 1) * 2)) {
                    return true
                }
            }
        }
    }


    /**
     * Удалить ключ из множества
     *
     * Алгоритм должен быть как минимум lock-free
     *
     * @param value значение ключа
     * @return false если ключ не был найден, true если ключ успешно удален
     */
    fun remove(value: T): Boolean {
        while (true) {
            val (pred, curr) = find(value)
            if (curr.data != value) {
                return false
            } else {
                val succ = curr.next.reference
                val currVersion = curr.version
                val predVersion = pred.version
                if (curr.next.compareAndSet(succ, succ, currVersion * 2, (currVersion + 1) * 2 + 1)) {
                    pred.next.compareAndSet(curr, succ, predVersion * 2, (predVersion + 1) * 2)
                    return true
                }
            }
        }
    }


    /**
     * Проверка наличия ключа в множестве
     *
     * Алгоритм должен быть как минимум wait-free
     *
     * @param value значение ключа
     * @return true если элемент содержится в множестве, иначе - false
     */
    operator fun contains(value: T): Boolean {
        var curr = head.next.reference
        while (true) {
            val succ = curr.next.reference
            if (curr.next.stamp % 2 == DELETED) {
                if (curr.data == value) {
                    return false
                }
                curr = succ
            } else {
                val data = curr.data
                if (data == value) {
                    return true
                }
                if (data == null || data > value) {
                    return false
                } else {
                    curr = succ
                }
            }
        }
    }

    /**
     * Возвращает lock-free итератор для множества
     *
     * @return новый экземпляр итератор для множества
     */
    operator fun iterator(): Iterator<T> {
        return makeSnapshot().listIterator()
    }

    fun print() {
        var curr = head.next.reference
        while (true) {
            val data = curr.data
            if (data == null) {
                print("--\n")
                return
            } else {
                print("--$data")
            }
            curr = curr.next.reference
        }
    }

    private fun find(value: T): Pair<Node<T>, Node<T>> {
        while (true) {
            var pred = head
            var curr = pred.next.reference
            while (true) {
                val succ = curr.next.reference
                if (curr.next.stamp % 2 == DELETED) {
                    val predVersion = pred.version
                    if (!pred.next.compareAndSet(curr, succ, predVersion * 2, predVersion * 2)) {
                        break
                    }
                    curr = succ
                } else {
                    val data = curr.data
                    if (data == null || data >= value) {
                        return Pair(pred, curr)
                    } else {
                        pred = curr
                        curr = succ
                    }
                }
            }

        }
    }

    private fun makeSnapshot(): List<T> {
        var first = unSafeSnapshot()
        while (true) {
            val second = unSafeSnapshot()
            if (first == second) {
                return second.map { it.second }
            } else {
                first = second
            }
        }
    }

    private fun unSafeSnapshot(): List<Pair<Int, T>> {
        val snapshot = mutableListOf<Pair<Int, T>>()
        var curr = head.next.reference
        while (true) {
            val data = curr.data
            if (data == null) {
                return snapshot
            } else if (curr.next.stamp % 2 == NOT_DELETED) {
                snapshot.add(Pair(curr.version, data))
            }
            curr = curr.next.reference
        }
    }


    private inner class Node<T>(nextNode: Node<T>?, val data: T?) {
        val next: AtomicStampedReference<Node<T>> = AtomicStampedReference<Node<T>>(nextNode,
            currentVersion.getAndIncrement() * 2)
        val version: Int
            get() = next.stamp / 2
    }


    companion object {
        const val DELETED = 1
        const val NOT_DELETED = 0
    }
}